core/src/main/resources/hudson/slaves/JNLPLauncher/config.jelly

```diff
@@ -28,6 +28,9 @@ THE SOFTWARE.
   <j:if test="${descriptor.workDirSupported}"> 
     <f:property title="${%Enable work directory}" field="workDirSettings" />
   </j:if>
+  <f:entry field="webSocket">
```

Comment:
This works a little differently than I expected. Or a little inconsistently.

A) If I don't enable WebSocket here:
  1) I can launch an agent normally and it connects via the TCP port.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

B) If I do enable WebSocket here:
  1) If I try to launch an agent normally, it gets the "-webSocket" argument in the jnlp file and connects via WebSocket.
  2) I can launch an agent with the "-webSocket" option and it connects via WebSocket.
  3) I can launch an agent with a direct connection, "-direct".

In other words, the agent can control if it wants to connect via WebSocket or direct but it can't control if it wants to connect via jnlp file to TCP port.

Things are better when they are regular and consistent. But I'm not certain this is really a problem here, that it needs to be changed, or what it would be changed to. 

Theoretically the config UI could have a choice to connect via i) config (jnlp) file, ii) direct, or iii) WebSocket and then the server would only allow that mechanism. 

Another possibility would be that the WebSocket implementation check connections and not allow them if they're not configured for WebSocket.

The inconsistent behavior had me confused for a while until I mapped it out but maybe it's not going to be confusing for users and developers.

Just recording my observations and musing about it.

Reply:
You are correct that the server does not attempt to reject connections made using a different transport than the selected one. (`-direct` is not a distinct option—this is still using the TCP port, just with the port number being discovered differently.) The checkbox controls whether the CLI argument `-webSocket` is offered to `Main` when `Launcher` is run with `-jnlpUrl`. In other words, the server merely uses this to serve up appropriate `/computer/xxx/slave-agent.jnlp` content.

Reply:
Remember that `Main` is the fundamental entry point for inbound agents. `Launcher` is the wrapper entry point that parses JNLP to call `Main`, as well as handling stdio mode for outbound agents (and a TCP _server_ mode which AFAIK is unused by any Jenkins `ComputerLauncher`). It would be nice if Javadoc and/or Markdown clarified the distinction, as it is hard to keep track of (see https://github.com/jenkinsci/docker-jnlp-slave/issues/87 for example though I think I am abandoning that).

Reply:
Yeah, I think the existing behavior is pretty reasonable the more I think about it. I wanted to record my observations and see if anything else popped out.